pkg/conversion/mapi2capi/aws.go

```diff
@@ -41,11 +41,26 @@ var (
 )
 
 const (
-	errInfrastructureInfrastructureNameCannotBeNil  = "infrastructure cannot be nil and infrastructure.Status.InfrastructureName cannot be empty"
-	errUnableToFindAMIReference                     = "unable to find a valid AMI resource reference"
-	errUnableToConvertUnsupportedAMIFilterReference = "unable to convert AMI Filters reference. Not supported in CAPI"
-	errUnableToConvertUnsupportedAMIARNReference    = "unable to convert AMI ARN reference. Not supported in CAPI"
-	errUnableToFindInstanceID                       = "unable to find InstanceID in ProviderID"
+	// Errors.
+	errInfrastructureInfrastructureNameCannotBeNil                   = "infrastructure cannot be nil and infrastructure.Status.InfrastructureName cannot be empty"
+	errUnableToFindAMIReference                                      = "unable to find a valid AMI resource reference"
+	errUnableToConvertUnsupportedAMIFilterReference                  = "unable to convert AMI Filters reference. Not supported in CAPI"
+	errUnableToConvertUnsupportedAMIARNReference                     = "unable to convert AMI ARN reference. Not supported in CAPI"
+	errUnableToFindInstanceID                                        = "unable to find InstanceID in ProviderID"
+	errLoadbalancersNotSupported                                     = "loadBalancers are not supported"
+	errNetworkInterfaceUnsupportedValue                              = "networkInterface type must be one of ENA or omitted, unsupported value"
+	errDeviceIndexMustBeZero                                         = "deviceIndex must be 0 or unset"
+	errMetadataNotSupported                                          = "metadata is not supported"
+	errUnsupportedAuthenticationValue                                = "unsupported authentication value"
+	errNonRootVolumesMustBeDeletedOnTerminationUnsupportedValueFalse = "non-root volumes must be deleted on termination, unsupported value false"
+	errVolumeSizeRequiredButMissing                                  = "volumeSize is required, but is missing"
+	errPlacementRegionShouldMatchInfrastructureStatusValue           = "placement.region should match infrastructure status value"
+	errVirtualNameNotSupported                                       = "virtualName is not supported"
+	errNoDeviceIsNotSupported                                        = "noDevice is not supported"
+
+	// Warnings.
+	warnRootVolumeMustBeDeletedOnTerminationIgnoringValueFalse = "root volume must be deleted on termination, ignoring invalid value false"
+	warnMissingBlockDeviceConf                                 = "missing ebs configuration for block device"
```

Comment:
I don't think extracting these out is the right move personally, I'm not sure it adds value and is making the errors harder to read personally, they aren't being re-used anywhere (other than in tests where I have a separate comment), I'd personally prefer these go back inline and that we match on full error strings rather than these substrings

Reply:
I tend to disagree, I dislike having to match things in text two times and having to amend both if the text needs changing.

I tried to make the var names as descriptive as possible so we shouldn't need to look up the whole text, every time we want to know what they throw.

Reply:
That's fair, lets see if you can find an alternative path in the middle somewhere

One benefit of typing it out twice though, is that I've picked up spelling mistakes that way many times 😅

Reply:
Done! Taken an approach that uses fmt.Sprintf with a typed out approach and then uses the variable as the last bit of the error. This mimics the [CPMS approach](https://github.com/openshift/cluster-control-plane-machine-set-operator/blob/8d85b2cb2c1dc8709fa49470b1df91640c642a59/pkg/controllers/controlplanemachineset/updates_test.go#L342).

Unfortunately I couldn't match certain warning strings composites because the warning string returned by the converter contains pointers that are not matchable from the testing side. So I kept those as matching only with the fixed string at the tail of the warning.

Reply:
Yeah it would yield errors like:

```
• [FAILED] [0.001 seconds]
mapi2capi AWS convert MAPI Machine [It] With VirtualName specified and missing EBS configuration
/Users/ddonati/src/cluster-capi-operator/pkg/conversion/mapi2capi/aws_test.go:247

  [FAILED] should match expected warnings while converting an AWS MAPI Machine to CAPI
  Expected
      <[]string | len:1, cap:1>: [
          "spec.providerSpec.value.blockDevices[0]: Invalid value: v1beta1.BlockDeviceMappingSpec{DeviceName:(*string)(nil), EBS:(*v1beta1.EBSBlockDeviceSpec)(nil), NoDevice:(*string)(nil), VirtualName:(*strin
g)(0xc000f85ea0)}: missing ebs configuration for block device",
      ]
  to consist of
      <[]*matchers.ContainSubstringMatcher | len:1, cap:1>: [
          {
              Substr: "spec.providerSpec.value.blockDevices[0]: Invalid value: v1beta1.BlockDeviceMappingSpec{DeviceName:(*string)(nil), EBS:(*v1beta1.EBSBlockDeviceSpec)(nil), NoDevice:(*string)(nil), Virtual
Name:(*string)(0xc000056d10)}: missing ebs configuration for block device",
              Args: nil,
          },
      ]
```
Where there is a pointer mismatch even if I declare them at the top like this:
```
	testValue                          = ptr.To[string]("test")
	blockDeviceMappingWithVirtualName  = &mapiv1.BlockDeviceMappingSpec{VirtualName: testValue}
	blockDeviceMappingWithoutEBSConfig = &mapiv1.BlockDeviceMappingSpec{DeviceName: ptr.To("/dev/sdb")}
```
And reference them in the inputs and in the expected output.

Reply:
But I have a new idea on how to circumvent this.
```
- warnings = append(warnings, field.Invalid(fldPath.Index(i), mapping, warnMissingBlockDeviceConf).Error())
+ warnings = append(warnings, field.Invalid(fldPath.Index(i).Child("ebs"), mapping.EBS, warnMissingBlockDeviceConf).Error())
```
Change the warning throw like so, so we can just match against nil

Reply:
WDYT?

Reply:
I'm surprised it is rendering the pointers, rather than rendering something more akin to a `%+v`, but, ok, lets not stress too much on this then

Reply:
I agree it is a bit overzelous in using `%#v`.
Ok then, I'll move to matching the lowest field to avoid the issue entirely. TY

Reply:
The more I think about this thread, the less I like extracting these.

Think about a case where the error message gets confused. Lets say I accidentally update one of these vars to mean something else, now, when its returned in the original location, I get a nonsense message that doesn't mean what I wanted it to.

I may think it's harmless, and it probably won't get picked up in code review, because we likely aren't looking at the context in which the variable is used.

Since the test uses the same variable, the test will still pass, and this will only get picked up when someone hits the error, and we find and get confused by the nonsense string.

When the errors are typed out inline as they were before, changing them is obvious and the diff will show the context -> less likely to produce a nonsense message.

And then add to that, I have to go and update a test. That is a conscious effort to change the test expectation and we will clearly see that we are changing the expected string. I'm not a fan of calculating test output. Calculated output can change underneath us without us noticing, we should write the strings out in full to ensure that if anything does change, the tests are either broken, or updated, and we can use them as a safety net.

Note the CPMS example you chose is different because it is wrapping a specific static error that is used only in testing as a sentinel error returned by the mocking library implementing an interface, and not an error returned by the controller itself.

Reply:
Had a chat offline, converged to inline errors. PR updated.